### ISSUE TITLE
Ruby pact provider ignoring regular expression (pact generated by pact-jvm)

### ISSUE THREAD

**Issue by agenteo** [realestate-com-au/pact#138]
Hi,

I've built a Spring Boot consumer (https://github.com/agenteo/spring-boot-pact-consumer) with pact-jvm and testing a Ruby app (https://github.com/agenteo/ruby-sinatra-pact-provider) with pact-provider.

If the fields are missing I get a test fail but the provider doesn't seem to be checking for regular expressions **that is on an array element**.

The generated pact file (https://github.com/agenteo/spring-boot-pact-consumer/blob/master/target/pacts/aggregator_app-seal_API.json) has this:

```
"$.body[*].status": {
                        "regex": "\\d*"
                    },
```

in the matching rules. You can see the sinatra app returning (https://github.com/agenteo/ruby-sinatra-pact-provider/blob/master/views/offers.json.erb) a static string of "Active" for status.

Before digging in to the source code I wanted to check if I missed some configuration. I read the provider is meant to pickup this is v2 of the spec automatically, did I miss something else?

Thanks

**Comment by uglyog** [comment]
This may be due to a defect in the pact jvm implementation, as it is missing the `"match": "regex"`

**Comment by bethesque** [comment]
Yeah, I think you're right Ron.

**Comment by bethesque** [comment]
Can you open an issue in pact-jvm please @agenteo?

**Comment by agenteo** [comment]
FYI I've tried to manually add that `"match": "regex"` to the pact file and the provider was still ignoring the regex. My gut feeling is something to do with the array [*] syntax since when I manually changed that to [1] the provider was correctly checking the regular expression. I can re-open on the other repo but wanted to bring up this additional information first.

**Comment by agenteo** [comment]
I just verified that having a missing `"match": "regex"` in a pact file with $.body[0] isn't a problem and correctly verified regular expressions. 
You can verify that using https://github.com/agenteo/ruby-sinatra-pact-provider and this pact file (https://pastebin.com/cAYNZENu).

Here's the pact with the [*] syntax https://pastebin.com/YKstLULQ where the Ruby provider ignores regular expressions on (without `"match": "regex"`). When I run `rake pact:verify` it says `WARN: Ignoring unsupported matching rules {"min"=>0} for path $['body']`. So I manually dropped that option and also added the match regex here https://pastebin.com/Lu6qNctS and now the provider is doing an exact match:

```
        [
          {
       -    "price": 2000.22,
       -    "offer_start_date": "2012-03-02T08:00:00.000Z",
       -    "destination_name": "Cancun",
       -    "status": "OK"
       +    "price": 2000.09,
       +    "offer_start_date": "2014-07-01T03:59:59.999Z",
       +    "destination_name": "Mantova",
       +    "status": "Ciao"
          }
        ]

       Key: - means "expected, but was not found".
            + means "actual, should not be found".
```

Is there a JSON spec of what a pact should look like? I was using the [*] version to match each array element. Is that what it's meant to do? I didn't find clear answer in the docs.

**Comment by uglyog** [comment]
There is no JSON spec atm for V2 matchers, the only available documentation is https://github.com/DiUS/pact-jvm/wiki/Matching#matching-rules

**Comment by uglyog** [comment]
Apart from the missing `match` attributes, it looks ok to me. @bethesque there may be an issue with the Ruby version too.

**Comment by bethesque** [comment]
Bugs aside, I'm not actually sure it will work - the status is a number, and we don't do regex on numbers.

**Comment by bethesque** [comment]
Oh, I beg your pardon, it's the body status, not the response status.

**Comment by bethesque** [comment]
I'm not sure if this is the cause or not, but the fact that it's saying `Ignoring unsupported matching rules {"min"=>0}` means it's missing the "$.body[*].status[*]*" rule, that already has an open issue (that I can't currently find).

**Comment by bethesque** [comment]
Here's the ruby test for regular expressions inside arrays. Note that the serialisation code is expected to write the rule `"$.body.alligators[*].*" => {"match" => "type"}`

```ruby
context "with an ArrayLike with a Pact::Term inside" do
          let(:matchable) do
            {
              body: {
                alligators: Pact::ArrayLike.new(
                  name: 'Fred',
                  phoneNumber: Pact::Term.new(generate: '1234567', matcher: /\d+/)
                )
              }
            }
          end

          let(:rules) do
            {
              "$.body.alligators" => {"min" => 1},
              "$.body.alligators[*].*" => {"match" => "type"},
              "$.body.alligators[*].phoneNumber" => {"match" => "regex", "regex" => "\\d+"}
            }
          end

          it "lists a rule that specifies that the regular expression must match" do
            expect(subject).to eq rules
          end
        end

```

**Comment by uglyog** [comment]
`status` is just an attribute, the problem may be that it is not handling an array at the root of the body. Hence the `"$.body": { "min": 0, "match": "type" }` rule.

**Comment by bethesque** [comment]
There is no JSON spec, but there is a test case in the pact-specification https://github.com/pact-foundation/pact-specification/blob/version-2/testcases/request/body/array%20with%20regular%20expression%20in%20element.json

**Comment by bethesque** [comment]
I'm trying to recreate the situation in ruby code, and I'm getting the error I would expect.

### Consumer

```ruby
      bar_service.
        upon_receiving("a retrieve thing request").with({
        method: :get,
        path: '/thing'
      }).
        will_respond_with({
        status: 200,
        headers: { 'Content-Type' => 'application/json' },
        body: Pact.each_like({status: Pact.term(/\d+/, "4")})
      })
```

### Pact

```json
{
  "consumer": {
    "name": "Foo"
  },
  "provider": {
    "name": "Bar"
  },
  "interactions": [
    {
      "description": "a retrieve thing request",
      "request": {
        "method": "get",
        "path": "/thing"
      },
      "response": {
        "status": 200,
        "headers": {
          "Content-Type": "application/json"
        },
        "body": [
          {
            "status": "4"
          }
        ],
        "matchingRules": {
          "$.body": {
            "min": 1
          },
          "$.body[*].*": {
            "match": "type"
          },
          "$.body[*].status": {
            "match": "regex",
            "regex": "\\d+"
          }
        }
      }
    }
  ],
  "metadata": {
    "pactSpecificationVersion": "2.0.0"
  }
}
```

### Provider
```ruby
class BarApp
  def call env
    [200, {'Content-Type' => 'application/json'}, [{"status" => "Active"},{"status" => "Active"}].to_json]
  end
end
```

<img width="827" alt="screen shot 2017-05-18 at 11 35 55 am" src="https://cloud.githubusercontent.com/assets/446228/26182856/39dae38e-3bbe-11e7-8962-b552e9303bcb.png">

**Comment by bethesque** [comment]
I'm also getting `WARN: Ignoring unsupported matching rules {"min"=>1} for path $['body']`, so something is not quite right though.

**Comment by uglyog** [comment]
I have released 3.3.10 of pact-jvm with a fix for matchers with missing "type" entries
